APIServer/appscale/api_server/log/messages.py

```diff
@@ -0,0 +1,292 @@
+            http_version: A string specifying the HTTP version.
+        """
+        self.request_id = request_id
+        self.project_id = project_id
+        self.service_id = service_id
+        self.version_id = version_id
+        self.ip = ip
+        self.nickname = nickname
+        self.user_agent = user_agent
+        self.host = host
+        self.method = method
+        self.resource = resource
+        self.http_version = http_version
+
+        self.status = None
+        self.response_size = None
+
+        # Unix timestamps in microseconds
+        self.start_time = int(time.time() * 1000 * 1000)
+        self.end_time = None
+
+        self.offset = None
+        self.app_logs = deque(maxlen=self.MAX_APP_LOGS)
+
+    @property
+    def combined(self):
+        """ A string representing the request as a combined log entry. """
+        timestamp = time.localtime(self.end_time / 1000 / 1000)
+        date = time.strftime('%d/%b/%Y:%H:%M:%S %z', timestamp)
+        return ('{ip} - {nickname} [{date}] "{method} {resource} HTTP/{ver}" '
+                '{status} {size} "{ua}"').format(
+            ip=self.ip, nickname=self.nickname, date=date, method=self.method,
+            resource=self.resource, ver=self.http_version, status=self.status,
+            size=self.response_size, ua=self.user_agent)
+
+    @staticmethod
+    def from_capnp(log):
+        """ Creates a RequestLog from a capnp RequestLog. """
+        service_id = 'default'
+        if ':' in log.versionId:
+            service_id, version_id = log.versionId.split(':')
+        else:
+            version_id = log.versionId
+
+        request_log = RequestLog(
+            log.requestId, log.appId, service_id, version_id, log.ip, log.nickname,
+            log.userAgent, log.host, log.method, log.resource, log.httpVersion)
+
+        request_log.status = log.status
+        request_log.response_size = log.responseSize
+        request_log.start_time = log.startTime
+        request_log.end_time = log.endTime
+        if log.offset:
+            request_log.offset = log.offset
+
+        request_log.app_logs.extend(
+            [AppLog.from_capnp(line) for line in log.appLogs])
+        return request_log
+
+    def to_capnp(self):
+        """ Creates a capnp RequestLog. """
+        request_log = logging_capnp.RequestLog.new_message()
+        for capnp_field in ['requestId', 'appId', 'ip', 'nickname', 'userAgent',
+                            'host', 'method', 'resource', 'httpVersion', 'status',
+                            'responseSize', 'offset', 'startTime', 'endTime']:
+            field = self.CAPNP_FIELDS.get(capnp_field, capnp_field)
+            value = getattr(self, field)
+            if value is not None:
+                setattr(request_log, capnp_field, value)
+
+        request_log.versionId = ':'.join([self.service_id, self.version_id])
+        app_logs = request_log.init('appLogs', len(self.app_logs))
```

Comment:
Is this app_logs variable ever appended to again? Because capnp lists initialized this way cannot be resized.

Reply:
No. All of the capnp serialization happens only when the message is sent.